pytorch_lightning/metrics/retrieval/retrieval_metric.py

```diff
@@ -0,0 +1,117 @@
+
+    query_without_relevant_docs_options = ['error', 'skip', 'pos', 'neg']
+
+    def __init__(
+        self,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+        query_without_relevant_docs: str = 'skip',
+        exclude: int = IGNORE_IDX
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn
+        )
+
+        if query_without_relevant_docs not in self.query_without_relevant_docs_options:
+            raise ValueError(
+                f"`query_without_relevant_docs` received a wrong value {query_without_relevant_docs}. "
+                f"Allowed values are {self.query_without_relevant_docs_options}"
+            )
+
+        self.query_without_relevant_docs = query_without_relevant_docs
+        self.exclude = exclude
+
+        self.add_state("idx", default=torch.tensor([], dtype=torch.int64), dist_reduce_fx="cat")
+        self.add_state("preds", default=torch.tensor([], dtype=torch.float32), dist_reduce_fx="cat")
+        self.add_state("target", default=torch.tensor([], dtype=torch.int64), dist_reduce_fx="cat")
+
+    def update(self, idx: torch.Tensor, preds: torch.Tensor, target: torch.Tensor) -> None:
+        if not (idx.shape == target.shape == preds.shape):
+            raise ValueError(
+                "Indexes, preds and targets must be of the same shape after preds normalization"
+            )
+        
+        idx = idx.to(dtype=torch.int64).flatten()
+        preds = preds.to(dtype=torch.float32).flatten()
+        target = target.to(dtype=torch.int64).flatten()
+
+        self.idx = torch.cat([self.idx, idx])
+        self.preds = torch.cat([self.preds, preds])
+        self.target = torch.cat([self.target, target])
+
+    def compute(self) -> torch.Tensor:
+        res = []
+        for group in get_mini_groups(self.idx):
+            if self.target[group].sum() == 0:
+                if self.query_without_relevant_docs == 'error':
+                    raise ValueError(
+                        f"`{self.__class__.__name__}.compute()` was provided with "
+                        f"a query without positive targets, indexes: {group}"
+                    )
+                if self.query_without_relevant_docs == 'pos':
+                    res.append(torch.tensor(1.0))
+                elif self.query_without_relevant_docs == 'neg':
+                    res.append(torch.tensor(0.0))
+            else:
+                res.append(
+                    self._metric(group)
+                )
+        return torch.stack(res).mean()
+
+    def _metric(self, group: List[int]) -> torch.Tensor:
+        r"""
+        Compute a metric over a single group.
+        """
+        raise NotImplementedError("This method must be overridden by subclasses")
```

Comment:
If the intention is that the user should not be using this class directly, but only subclass it, then it should be an abstract class and this should be an abstract method.